src/pow_fpga_accel.c

```diff
@@ -95,119 +82,69 @@ static bool PoWFPGAAccel(void *pow_ctx)
-        ctx[i].ctrl_fd = open(DEV_CTRL_FPGA, O_RDWR);
-        if (ctx[i].ctrl_fd < 0) {
-            perror("cpow-ctrl open fail");
-            goto fail_to_open_ctrl;
-        }
-        ctx[i].in_fd = open(DEV_IDATA_FPGA, O_RDWR);
-        if (ctx[i].in_fd < 0) {
-            perror("cpow-idata open fail");
-            goto fail_to_open_idata;
-        }
-        ctx[i].out_fd = open(DEV_ODATA_FPGA, O_RDWR);
-        if (ctx[i].out_fd < 0) {
-            perror("cpow-odata open fail");
-            goto fail_to_open_odata;
-        }
-        impl_ctx->bitmap = impl_ctx->bitmap << 1 | 0x1;
+    ctx->ctrl_fd = open(DEV_CTRL_FPGA, O_RDWR);
+    if (ctx->ctrl_fd < 0) {
+        perror("cpow-ctrl open fail");
+        goto fail_to_open_ctrl;
     }
-    impl_ctx->context = ctx;
-    pthread_mutex_init(&impl_ctx->lock, NULL);
-
-    devmem_fd = open("/dev/mem", O_RDWR | O_SYNC);
-    if (devmem_fd < 0) {
-        perror("devmem open");
-        goto fail_to_open_memopen;
+    ctx->in_fd = open(DEV_IDATA_FPGA, O_RDWR);
+    if (ctx->in_fd < 0) {
+        perror("cpow-idata open fail");
+        goto fail_to_open_idata;
     }
-
-    fpga_regs_map =
-        (uint32_t *) mmap(NULL, HPS_TO_FPGA_SPAN, PROT_READ | PROT_WRITE,
-                          MAP_SHARED, devmem_fd, HPS_TO_FPGA_BASE);
-    if (fpga_regs_map == MAP_FAILED) {
-        perror("devmem mmap");
-        goto fail_to_open_memmap;
+    ctx->out_fd = open(DEV_ODATA_FPGA, O_RDWR);
+    if (ctx->out_fd < 0) {
+        perror("cpow-odata open fail");
+        goto fail_to_open_odata;
     }
 
-    cpow_map = (uint32_t *) (fpga_regs_map + CPOW_BASE);
+    impl_ctx->context = ctx;
+    pthread_mutex_init(&impl_ctx->lock, NULL);
 
     return true;
 
-fail_to_open_memmap:
-    close(devmem_fd);
-fail_to_open_memopen:
-    close(ctx[i].out_fd);
 fail_to_open_odata:
-    close(ctx[i].in_fd);
+    close(ctx->in_fd);
 fail_to_open_idata:
-    close(ctx[i].ctrl_fd);
+    close(ctx->ctrl_fd);
 fail_to_open_ctrl:
 fail_to_malloc:
-    for (int j = i - 1; j > 0; j--) {
-        close(ctx[j].in_fd);
-        close(ctx[j].out_fd);
-        close(ctx[j].ctrl_fd);
-    }
     return false;
 }
 
 static void PoWFPGAAccel_Context_Destroy(ImplContext *impl_ctx)
 {
     PoW_FPGA_Accel_Context *ctx = (PoW_FPGA_Accel_Context *) impl_ctx->context;
-    for (int i = 0; i < impl_ctx->num_max_thread; i++) {
-        close(ctx[i].in_fd);
-        close(ctx[i].out_fd);
-        close(ctx[i].ctrl_fd);
-    }
-    free(ctx);
 
-    int result = munmap(fpga_regs_map, HPS_TO_FPGA_SPAN);
-    if (result < 0) {
-        perror("devmem munmap");
-    }
+    close(ctx->in_fd);
+    close(ctx->out_fd);
+    close(ctx->ctrl_fd);
 
-    close(devmem_fd);
+    free(ctx);
 }
 
 static void *PoWFPGAAccel_getPoWContext(ImplContext *impl_ctx,
                                         int8_t *trytes,
                                         int mwm)
 {
     pthread_mutex_lock(&impl_ctx->lock);
```

Comment:
Actually, this statement also can be dropped if you are sure that only 1 fpga board would be supported on the platform. Or it should be reserved for extensibility as well as `PoWFPGAAccel_freePoWContext`.

Reply:
Yes, I agree your point. I will drop this lock.